src/main/cpp/src/mixed_sort_merge_join.cu

```diff
@@ -0,0 +1,759 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mixed_sort_merge_join.hpp"
+
+#include <cudf/ast/detail/expression_evaluator.cuh>
+#include <cudf/ast/detail/expression_parser.hpp>
+#include <cudf/ast/expressions.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/join/sort_merge_join.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda_runtime_api.h>
+#include <thrust/copy.h>
+#include <thrust/count.h>
+#include <thrust/distance.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/remove.h>
+#include <thrust/sort.h>
+#include <thrust/tuple.h>
+#include <thrust/unique.h>
+
+namespace spark_rapids_jni {
+
+namespace {
+
+// =============================================================================
+// HELPER FUNCTIONS (Used by both conditional and non-conditional joins)
+// =============================================================================
+template <bool has_nulls>
+__global__ void filter_join_indices_kernel(
+  cudf::size_type const* __restrict__ left_indices,
+  cudf::size_type const* __restrict__ right_indices,
+  cudf::size_type num_pairs,
+  cudf::table_device_view left_table,
+  cudf::table_device_view right_table,
+  cudf::ast::detail::expression_device_view device_expression_data,
+  bool* __restrict__ keep_mask)
+{
+  extern __shared__ char raw_intermediate_storage[];
+  cudf::ast::detail::IntermediateDataType<has_nulls>* intermediate_storage =
```

Comment:
Since we're playing pointer games here with dynamic shared memory, I think it would be a good idea to make `cudf::ast::detail::IntermediateDataType<has_nulls>` a typedef to protect against discrepancies with the kernel call site.